ByteBuffer.writeLengthPrefixed

dnadoba opened the pull request:
A convenience method to write a message and prepend its length in front of it.

### Motivation:
Many protocols use a length field to split a byte stream into multiple separate messages. `swift-nio-extras` supports these kind of protocols through `LengthFieldPrepender` and friends. However, sometimes a channel handler is not the right tool, especially if the a message itself contains multiple dynamic length fields. One example is [RSocket Routing Metadata Extensions](https://github.com/rsocket/rsocket/blob/master/Extensions/Routing.md). 

A `ByteBuffer` is probably a better tool to create this message. This can be tricky to get right if strings are part of the message. A naive solution like the following has two issues:
```swift
func add(tag: String, to buffer: inout ByteBuffer) {
    let length = tag.count
    buffer.writeInteger(UInt8(length))
    buffer.writeString(tag)
}
```
The `count` property in Swift is different than in most other languages. It does **not** return the number of UTF-8 code points and can result in incorrect behavior if multi-byte characters are used (e.g. Emojis). Furthermore, the size of the string could be larger than `UInt8` can represent.
### Modifications:

This PR introduces a new method on `ByteBuffer` called `writeLengthPrefix(endianness:as:writeMessage:)`. It solves both problems mentioned above. It first reserves space for the length prefix and delegates the actually writing of the message to `writeMessage`. Afterwards, it counts how many bytes were written during the call to `writeMessage` and then sets it on the previously reserved space in front of the actually message.

### Result:
We can now write arbitrary content to a buffer and prefix it with the number of bytes written.
```swift
func add(tag: String, to buffer: inout ByteBuffer) {
    buffer.writeLengthPrefix(as: UInt8.self) { buffer in
        buffer.writeString(tag)
    }
}
```

### Alternatives
We could also name the method `prependLengthToMessage(endianness:as:writeMessage:)` but this would be against the general rule that all methods which increment the `writtenBytes` property should be prefixed with `write`.

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,55 @@
+    @inlinable
+    public mutating func writeLengthPrefix<Integer: FixedWidthInteger>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type = Integer.self,
```
I don't think we want the default here, the caller should explicitly provide the length type.

dnadoba replied:
Agree, changed 👍

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,55 @@
+        endianness: Endianness = .big,
+        as integer: Integer.Type = Integer.self,
+        message writeMessage: (inout ByteBuffer) throws -> ()
+    ) throws -> Int {
```
nit: the normal NIO style is to always use explicit `self`, we're missing a few in the body of this.

dnadoba replied:
Done ✅

glbrntt commented on Tests/NIOTests/ByteBufferLengthPrefixTests+XCTest.swift:
```diff
@@ -0,0 +1,40 @@
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 0 Apple Inc. and the SwiftNIO project authors
```
Something's gone wrong here with the year... was it generated like this?

Lukasa replied:
#1903.

dnadoba replied:
Good Catch, thanks. Updated to 2021

dnadoba commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,55 @@
+    /// - Returns: Number of total bytes written
+    @discardableResult
+    @inlinable
+    public mutating func writeLengthPrefix<Integer: FixedWidthInteger>(
```
`set` method added and renamed `write` method to `writeLengthPrefixed`.

I'm not sure about the `get`/`read` methods. They are fairly trivial to implement and there are no pitfalls that I am aware of at least. However, if you think they add value I can still add them.

Lukasa replied:
I think the pitfall is that it's easy to not enforce the read-length: that is, to issue a read and not police that a) the entire body is consumed, and b) that nothing is overconsumed.

dnadoba replied:
Okay, but I think a `get` method would **not** make sense because we could not police that or can we? However, I will add a `read` method.

dnadoba commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,55 @@
+        
+        let messageStartIndex = writerIndex
+        try writeMessage(&self)
+        let messageEndIndex = writerIndex
```
I have changed the `writeMessage` function to return a number of bytes written for now. The `set` version of this method actually requires it, otherwise we do not know how many bytes were actually written if only `set` methods are used during `writeMessage`.

However, I find it more convenient to not return the number of bytes written because it can become rather difficult to provide this number if you use any kind of control flow during `writeMessage`. I would like that `setLengthPrefixed` requires a number of bytes written to be returned but `writeLengthPrefixed ` should not require it and compute the bytes written through the indices. WDYT?

As an alternative, we could also require the `writeMessage` function of `writeLengthPrefixed` to return the number of bytes written and add an assert/precondition that this number is actually correct but we should definitely document this behavior.

Lukasa replied:
I think I'd rather keep them the same. In principle it shouldn't be too challenging to do this with control flow: a number of NIO projects including NIOSSH effectively include this function and manage their control flow fairly straightforwardly. The pattern tends to be:

```swift
buffer.writeLengthPrefixed(as: UInt32.self) { buffer in
    var written = 0
    written += buffer.writeString(self.someString)
    if self.foo {
       written += buffer.writeInteger(self.bar)
   else {
       written += buffer.writeInteger(self.baz)
    }
    return written
}

```

dnadoba replied:
Okay, let's keep it this way.

Just out of curiosity, do we get any benefit compared to the "automatic" calculation ?

Lukasa replied:
Yeah, better composition. When the inner function body returns `Void` then you can't write:

```swift
buffer.writeLengthPrefixed(as: UInt32.self) { $0.writeInteger(someInteger) }
```

With this version, you can. It also saves us a bunch of work, computationally. I'm more than happy for us to _police_ the bounds though: we can add an `assert` statement that enforces that the amount the user claims they wrote is actually what they wrote.

dnadoba replied:
You can actually write it like in your example because all write methods are marked with `@discardableResult`.

It should also only be two memory reads and one subtraction compared to the manual calculations. 
Maybe my intuitions are wrong but I thought that this should not be noticeable because whatever `writeMessage` does is probably be more expensive.

Lukasa replied:
Your intuitions are right, it's not that it's computationally expensive, it's that it's cognitively expensive. The load is simply more complex than allowing the method to say how many bytes it wrote.

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,64 @@
+    /// - Returns: Number of total bytes written i.e. size of the integer plus the number returned by `setMessage`
+    @discardableResult
+    @inlinable
+    public mutating func setLengthPrefixed<Integer: FixedWidthInteger>(
```
The semantics of this don't seem quite right to me, imagine:

```swift
var buffer = ByteBuffer()
buffer.writeBytes([0, 0, 0, 0])

buffer.setLengthPrefixed(at: self.readerIndex, as: UInt8.self) { buffer in
  buffer.writeBytes([1, 2, 3, 4])
}
```

`buffer` would be `[4, 0, 0, 0, 1, 2, 3, 4]` which would surprise people, I think.

I wonder if `setMessage` should also be given the index after the length bytes, i.e. the index where it should start setting bytes?

Lukasa replied:
Oh good note: I seem to recall having tripped over this elsewhere. I think the `set` variant is probably not sensible on length-prefixed writes for exactly this reason. Sorry about this suggestion @dnadoba, I'm wondering if we should just revert to only having a `write` version.

glbrntt replied:
Yeah, even if given an index to use in `setMessage` it's still pretty easy to get wrong.

dnadoba replied:
Agree, I will remove the `set` method again.

Lukasa commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,107 @@
+    public mutating func readLengthPrefixed<Integer, Result>(
+        endianness: Endianness = .big,
+        as integer: Integer.Type,
+        readMessage: (Integer, inout ByteBuffer) throws -> Result?
```
Ok, we need to be very careful with this API, I think.

Concretely, under what circumstances should `readMessage` return `nil`? We can only call the `readMessage` function if we have an appropriate number of bytes. `ByteBuffer` traditionally uses `nil` to signal "not enough bytes to parse", but if we call `readMessage` we got as many bytes as we're ever gonna get.

I think if `readMessage` returns `nil` we need to `throw`. This helps make the `Result?` optional unambiguous: it returns `nil` only if there weren't enough bytes for the full body to be present.

dnadoba replied:
I have changed the implementation and it now no longer throws at all. Reading too much is not possible because we pass `readMessage` a `slice` now. Not consuming all bytes is still possible and does not throw. However, I think we want to actually allow this because protocols like [RTP use padding bytes at the end of a message which should be ignored](https://datatracker.ietf.org/doc/html/rfc3550#page-37).

I think it is okay to return nil if `readMessage` returns nil. This mirrors the behavior of `readString(length:)` which also returns nil if decoding fails. Otherwise the following example would throw if String decoding fails:
```swift
buffer.readLengthPrefixed(as: UInt8.self) { length, buffer in 
    buffer.readString(length)
}
```
This would be surprising at least for me.

Lukasa replied:
`readString(length:)` does not return `nil` if decoding fails because decoding cannot fail. Specifically, we use [`String(decoding:as:)`](https://developer.apple.com/documentation/swift/string/2907004-init) which is non-failable. This is not the same as [`String(bytes:encoding:)`](https://developer.apple.com/documentation/swift/string/3126734-init) which _is_ failable.

There should be no `read` method in NIO that returns `nil` in a circumstance other than "there were not enough readable bytes to satisfy the operation", and if there is that's a bug.

dnadoba replied:
Ah I see! Thanks for taking the time to explain it to me :)

Lukasa commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,107 @@
+            return nil
+        }
+        guard messageLength >= 0,
+              messageLength <= readableBytes else {
```
```suggestion
              messageLength <= self.readableBytes else {
```

Lukasa commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,107 @@
+        }
+        
+        let readStartIndex = self.readerIndex
+        guard let result = try readMessage(messageLength, &self) else {
```
I think if we should `slice` ourselves here and pass the slice into the `readMessage` function. This will avoid the need to dance with indices, and will also let the various `readX` functions behave sensibly in the body of `readMessage`.

glbrntt replied:
+1 on passing in a slice. I think a separate `readLengthPrefixedSlice` would actually be a useful function to have (it could then be used as the base implementation for this function).

dnadoba replied:
Changed to use a `slice`. 
@glbrntt do you mean something like this:
```swift
mutating func readLengthPrefixedSlice<Integer>(
    endianness: Endianness = .big,
    as integer: Integer.Type
) -> ByteBuffer? where Integer: FixedWidthInteger {
    readLengthPrefixed(endianness: endianness, as: Integer.self) { _, slice in slice }
}
```

glbrntt replied:
More or less -- just the other way around:

```swift
public mutating func readLengthPrefixed<Integer, Result>(
    endianness: Endianness = .big,
    as integer: Integer.Type,
    readMessage: (Int, inout ByteBuffer) throws -> Result?
) rethrows -> Result? where Integer: FixedWidthInteger {
    var slice = self.readLengthPrefixedSlice(...)
    return try readMessage(slice.readableBytes, &slice)
}
```

dnadoba replied:
Added ✅

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,107 @@
+        }
+        guard messageLength >= 0,
+              messageLength <= readableBytes else {
+            return nil
```
By returning here we can leave the buffer in an odd state: we've read the message length but don't have enough bytes to read the message. If we subsequently append more bytes and call this function again we'll have lost the length from the buffer.

dnadoba replied:
I have changed the implementation to only move the `readerIndex` forward if everything was read successful.

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,107 @@
+        endianness: Endianness = .big,
+        as integer: Integer.Type,
+        writeMessage: (inout ByteBuffer) throws -> Int
+    ) throws -> Int  where Integer: FixedWidthInteger {
```
```suggestion
    ) throws -> Int where Integer: FixedWidthInteger {
```

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,107 @@
+    /// - Parameters:
+    ///     - endianness: The endianness of the length prefix `Integer` in this `ByteBuffer` (defaults to big endian).
+    ///     - integer: the desired `Integer` type used to write the length prefix
+    ///     - writeMessage:
```
nit: missing docs here

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,105 @@
+            return nil
+        }
+        
+        _moveReaderIndex(forwardBy: prefixSize + messageLength)
```
```suggestion
        self._moveReaderIndex(forwardBy: prefixSize + messageLength)
```

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,105 @@
+    ) rethrows -> Result? where Integer: FixedWidthInteger {
+        let prefixSize = MemoryLayout<Integer>.size
+        
+        guard let lengthPrefix = self.getInteger(at: readerIndex, endianness: endianness, as: Integer.self),
```
```suggestion
        guard let lengthPrefix = self.getInteger(at: self.readerIndex, endianness: endianness, as: Integer.self),
```

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,105 @@
+        
+        guard let lengthPrefix = self.getInteger(at: readerIndex, endianness: endianness, as: Integer.self),
+              let messageLength = Int(exactly: lengthPrefix),
+              var messageBuffer = self.getSlice(at: readerIndex + prefixSize, length: messageLength)
```
```suggestion
              var messageBuffer = self.getSlice(at: self.readerIndex + prefixSize, length: messageLength)
```

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,105 @@
+            return nil
+        }
+        
+        guard let result = try readMessage(messageLength, &messageBuffer) else {
```
Do we want to pass in the message length? It can be recovered from the slice if necessary.

glbrntt replied:
Also wondering if we should drop the `inout` when passing in the slice, I think it makes it clearer that "this is your slice, do what you want with it".

dnadoba replied:
The "Problem" is that you would need to make your own local copy to actually read something from that slice.
This would no longer work:
```swift
buffer.readLengthPrefixed(as: UInt8.self) { buffer in 
    buffer.readString(buffer.readableBytes)
}
```
You would need to write the following:
```swift
buffer.readLengthPrefixed(as: UInt8.self) { buffer in 
    var buffer = buffer
    return buffer.readString(buffer.readableBytes)
}
```

I'm also thinking about removing the `readLengthPrefixed` method and just keeping `readLengthPrefixedSlice` because you could just write something like this:
```swift
guard var slice = buffer.readLengthPrefixedSlice(as: UInt8.self),
           let message = slice.readString(slice.readableBytes) 
else {
    throw MyCustomError()
}
```
At least in the case of RSocket we probably want to throw a custom error anyway. WDYT?

Lukasa replied:
I think I'd be fine with that.

glbrntt replied:
Same here.

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,105 @@
+            return nil
+        }
+        
+        _moveReaderIndex(forwardBy: prefixSize + messageLength)
```
We need to the reader index forward if `readMessage` throws or returns `nil` (since we read enough bytes from the buffer).

dnadoba replied:
Done ✅

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,105 @@
+        // Write a zero as a placeholder which will later be overwritten by the actual number of bytes written
+        totalBytesWritten += self.writeInteger(.zero, endianness: endianness, as: Integer.self)
+        
+        let startWriterIndex = writerIndex
```
```suggestion
        let startWriterIndex = self.writerIndex
```

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,105 @@
+        
+        let startWriterIndex = writerIndex
+        let messageLength = try writeMessage(&self)
+        let endWriterIndex = writerIndex
```
```suggestion
        let endWriterIndex = self.writerIndex
```

glbrntt commented on Sources/NIOCore/ByteBuffer-lengthPrefix.swift:
```diff
@@ -0,0 +1,144 @@
+            throw LengthPrefixError.messageCouldNotBeReadSuccessfully
+        }
+        return result
+    }
+    /// Reads an `Integer` from `self` and reads a slice of that length from `self`
```
```suggestion
    }
    
    /// Reads an `Integer` from `self` and reads a slice of that length from `self`
```

glbrntt commented on Tests/NIOTests/ByteBufferLengthPrefixTests.swift:
```diff
@@ -0,0 +1,230 @@
+        )
+    }
+    func testReadMessageWithMaliciousLength() {
+        buffer.writeInteger(UInt64(Int.max) + 1)
```
Can we have a similar test with a negative length?

glbrntt reviewed (commented):
I left a formatting suggestion and I think we need one more test. Otherwise this looks great. Up to you whether you want to leave in `readLengthPrefixed` or not.

dnadoba commented:
I added a space between method declarations, added a test for a negative length prefix and updated the documentation.
I decided to keep the `readLengthPrefixed` method to be more in line with the `writeLengthPrefixed` method.

glbrntt reviewed (approved):
Looks great, thanks @dnadoba!

Lukasa commented:
@dnadoba Can I recommend you build RSocket with NIO `main`? I believe you'll find that this patch has introduced an ambiguity with your downstream code.

dnadoba commented:
Yes it has. `ByteBuffer.LengthPrefixError` was also defined by RSocketCore. Thank you! Fixed by https://github.com/rsocket/rsocket-swift/pull/57
